Initial wrapper implementation

jonseitz opened the pull request:
Kind of breaking form by doing this initial implementation in `develop`, but given that it's a very small library I don't think that's a _huge_ issue.

This sets up our `OracleWrapper` class, which is based on the `OracleDB` class that's defined as part of the EAF ETL process, with some small changes aimed at making the class more generalizable. It's also in Typescript now too.

The tests are all pretty basic unit tests. I had spent some time trying to get a oracle database container running for more invovled integration tests but there's just not a great solution for doing that quickly.

jonseitz committed:
Fix comma-dangle issue

There's a weird conflict between the eslint rule and the
typescript-eslint rule re: comma-dangle, so we need to make sure it's
defined the same for both.

jonseitz committed:
Create Oracle Wrapper

Defines our OracleWrapper class, which just simplifies the process of
creating connection pools and running large queries against the
database.

jonseitz committed:
Namespace package under @seas-computing

Since we're the primary audience for this, we'll add the @seas-computing
namespace for publishing on npm.

jonseitz committed:
Updates typedoc to fix vulnerability

Installs the latest version of typedoc, which requires some changes to the
config file options.

jonseitz committed:
Merge branch 'main' into develop

rmainwork commented on src/index.ts:
```diff
@@ -0,0 +1,147 @@
+ */
+
+export default class OracleWrapper {
+  /**
+   * A cache alias that will be used to identify this connection
+   * defaults to using the sid from the credentials
+   */
```
That's a nice feature!

rmainwork commented on src/index.ts:
```diff
@@ -0,0 +1,147 @@
+  public async releasePool(): Promise<void> {
+    if (this.connectionPool) {
+      try {
+        await this.connectionPool.close(60);
```
Might it be worth having `60` as a configurable property in the constructor (or as a class constant)? Rather than a magic number in the method like this...

jonseitz replied:
Good point. I made all the magic numbers configurable in 01f8ff1

rmainwork commented on src/index.ts:
```diff
@@ -0,0 +1,147 @@
+        this.connectionPool = null;
+        this.logger.info('Connection pool released');
+      } catch (err) {
+        this.logger.error('Could not release pool');
```
Might be worth also logging `err` in here for debugging?

jonseitz replied:
Good idea. Added in 3c26850.

rmainwork commented on src/index.ts:
```diff
@@ -0,0 +1,147 @@
+    try {
+      result = await conn.execute(query, variables, options);
+    } catch (err) {
+      this.logger.error('Query failed');
```
Why not include `err` here for debugging?

jonseitz replied:
Added in 3c26850.

rmainwork commented on src/index.ts:
```diff
@@ -0,0 +1,147 @@
+      await conn.release();
+      return response;
+    } catch (err) {
+      this.logger.error('Error fetching data from query');
```
As above regarding `err`

jonseitz replied:
Added in 3c26850.

rmainwork commented on .eslintrc.json:
```diff
@@ -18,6 +18,16 @@
     "@typescript-eslint"
   ],
   "rules": {
+    "@typescript-eslint/comma-dangle": [
```
You mentioned in our commit message that this is the fix for a conflict between two rules. Might this be better fixed upstream in `seas-computing/boilerplate-library`?

jonseitz replied:
Ah, right. I think it would need to be updated in both places, actually. I'll make that in the seas-computing/boilerplate-library repo as well.

rmainwork commented on src/index.ts:
```diff
@@ -0,0 +1,147 @@
+    // Iterate over response rows, 1000 at a time, creating a complete response
+    // https://jsao.io/2015/07/an-overview-of-result-sets-in-the-nodejs-driver/
+    try {
+      const fetchAndConcat = async (prevRows: T[]): Promise<T[]> => {
```
Never seen recursion in the wild before, cool!

rmainwork commented on src/__tests__/index.test.ts:
```diff
@@ -0,0 +1,512 @@
+  let connectionPool: Record<string, SinonStub>;
+  let dbConnection: Record<string, SinonStub>;
+  let queryResult: Record<string, Record<string, SinonStub>>;
+  const credentials = {
```
Should this be in a separate test data file somewhere?

jonseitz replied:
Yeah, that makes sense. My original thinking was that it wasn't really necessary for such a small library, but it's better to be consistent with our other projects and that will make it easier to expand in the future. Done in  6a7ade2

rmainwork commented on src/__tests__/index.test.ts:
```diff
@@ -0,0 +1,512 @@
+import util from 'util';
+import assert from 'assert';
```
Is there a reason for importing `assert` like this, instead of `import { strictEqual } from 'assert'`?

jonseitz replied:
Not really. I switched to individual methods in  86791db, for consistency.

rmainwork commented on src/__tests__/index.test.ts:
```diff
@@ -0,0 +1,512 @@
+  let testLogger: Record<string, SinonStub>;
+
+  beforeEach(function () {
+    testLogger = {
+      log: stub(),
+      error: stub(),
+      warn: stub(),
+      info: stub(),
+      debug: stub(),
+      verbose: stub(),
+    };
+    connectionPool = {
+      close: stub(),
+    };
+    dbConnection = {
+      execute: stub(),
+      release: stub(),
+    };
+    queryResult = {
+      resultSet: {
+        getRows: stub(),
+        close: stub(),
+      },
+    };
+    options = {
+      alias: 'TestDB1',
+      logger: testLogger as unknown as Logger,
+    };
```
Should these mocks be in their own `__mocks__` folder?

jonseitz replied:
I don't think we could split these out, since we need the mocks to be instantiated within the scope of mocha, so that they'll be reset as expected after each test.

rmainwork replied:
Ah right. Jest lets you do that - for some reason I thought Mocha did too.

rmainwork commented on src/__tests__/index.test.ts:
```diff
@@ -0,0 +1,512 @@
+        });
+        it('Should use the alias as the property', function () {
+          // eslint-disable-next-line
+          assert.strictEqual(aliasedWrapper['alias'], testAlias);
+        });
+        it('Should use the alias in the poolCredentials', function () {
+          // eslint-disable-next-line
+          assert.strictEqual(aliasedWrapper['poolCredentials'].poolAlias, testAlias);
```
There are a few instances of `eslint-disable-next-line`. Might it be better to either use something more targeted like `// eslint-disable-next-line @typescript-eslint/dot-notation` or update the ruleset to allow for private property access like this?

jonseitz replied:
Good point. I narrowed that down in eb4b8de. I'd rather not tweak the ruleset for this; it's really just a workaround to access the private properties and I think it's something we'd want to be deliberate about.

rmainwork commented on src/__tests__/index.test.ts:
```diff
@@ -0,0 +1,512 @@
+    let testQueryWithoutVariables: string;
+    let testVariables: Record<string, string>;
+    beforeEach(function () {
+      testError = new Error('query failed');
+      testQueryWithoutVariables = 'SELECT * FROM USERS;';
+      testQuery = "SELECT * FROM USERS WHERE NAME=':name';";
+      testVariables = { name: 'foo' };
```
Again, maybe these should all be in a test data file?

jonseitz replied:
Done as part of 6a7ade2

rmainwork reviewed (changes requested):
Overall, this looks great. Just a few small issues. Looking forward to having this going forward!

jonseitz committed:
Allow overriding of more options

The original implementation did not allow configuration of some of the
connection options, but there is a reasonable case for making more of
them configurable to accomodate different circumstances.

jonseitz committed:
Log errors in catch blocks

In addition to logging a message about the error, we can also log the
full error object.

jonseitz committed:
Narrow use of eslint-disable

We only really need to diable the dot-notation rule in order to access
private properties with bracket notation, so we should specify that in
the eslint-disable-next-line directive

jonseitz committed:
Don't test unnecessary things

Since we're now merging the default options and user options, we don't
really need to check the conditional assignment. That would just be
testing that the `...` operator  works.

Additionally, we don't really need to test that the alias is being added
to the poolCredentials property, and there's no point in having
pointless tests.

jonseitz committed:
Move test data into its own file

This is more in line with our style, and generally streamlines the
tests by cutting down the amount of data instantiation that needs to
happen.

jonseitz committed:
Import individual assert methods

This is for consistency with our other projects.

jonseitz committed:
Provide message with fail() call

Provides some additional context for why those tests would be failing,
if they failed

rmainwork commented on src/__tests__/mockData.ts:
```diff
@@ -0,0 +1,62 @@
+
+/**
+ * The fake results from our fake query
+ */
```
Minor thing I just spotted but, does this need a blank line between it and the thing it's documenting for consistency with the rest of the docblocks?

jonseitz replied:
Yeah, that's all over the place isn't it? I made it 1 space above and 0 below across the board in fe5ebea.

rmainwork replied:
lol, it's not just here - it's like that in some of our other projects I think too - just wanted to point it out while it was relatively small :)

rmainwork commented on src/__tests__/mockData.ts:
```diff
@@ -0,0 +1,62 @@
+};
+
+/**
+ * A fake query without an parameters
```
```suggestion
 * A fake query without parameters
```

jonseitz replied:
D'oh. Corrected that to **any** in d7d4759

rmainwork commented on src/__tests__/index.test.ts:
```diff
@@ -0,0 +1,481 @@
+            strictEqual(testLogger.error.args[1][0], testError);
+          }
+        });
+        it('Should throw the error', async function () {
```
~~Should we also test in here (or in another test case) that the error thrown is logged to `logger.error`?~~

rmainwork replied:
Ignore that. I somehow missed the test case above :clown_face:

rmainwork reviewed (changes requested):
Looks great, still a couple more things.

jonseitz committed:
Make docblock spacing consistent

The spacing for documentation was all over the place, but the most
consistent inconsistency seemed to be

1.) blank line above the docblock
2.) no blank line between the docblock and the thing being documented

jonseitz committed:
Correct typo
